{
 "metadata": {
  "name": "",
  "signature": "sha256:6ff02ed224ff58d3bdc8af24c31e7957e30315d19007df352c9ff97e1d0f5e4c"
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "from Bio import SeqIO\n",
      "\n",
      "import os\n",
      "\n",
      "#try to avoid using OS, and open python notebook from the dir that the DB will be in (ie. C:\\Program Files\\NCBI\\Blast-2.2.29+\\db\\extract)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 10
    },
    {
     "cell_type": "raw",
     "metadata": {},
     "source": [
      "#print os.getcwd()"
     ]
    },
    {
     "cell_type": "raw",
     "metadata": {},
     "source": [
      "#os.chdir(\"C:\\Program Files\\NCBI\\Blast-2.2.29+\\db\\extract\")\n",
      "\n",
      "fasta = \"plantall.fna\"\n",
      "\n",
      "countrec = 0\n",
      "for seq_rec in SeqIO.parse(fasta, \"fasta\"):\n",
      "    countrec +=1\n",
      "print countrec\n",
      "\n",
      "#!makeblastdb -in plantall.fna -dbtype nucl -parse_seqids -out refseqplant\n",
      "\n",
      "#execute through terminal window ! (without !)\n"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "#HAVE IT WRITE A FILE PER ORGANISM\n",
      "#inputfile = \"QGG24G02.yg.ab1_CLS_S3_Contig10858_LettuceNCBI.fasta\"\n",
      "#outputfile = inputfile+\"_\"+\"refseqBLASTresults.csv\"\n",
      "                   \n",
      "inputfile = \"C:\\Users\\Alex\\Desktop\\Work\\Research\\Lettuce\\QGG24G02.yg.ab1_CLS_S3_Contig10858_LettuceNCBI.fasta\"\n",
      "outputfile = \"C:\\Users\\Alex\\Desktop\\Work\\Research\\Lettuce\\\"\n",
      "\n",
      "blastn_cline = NcbiblastnCommandline(query=inputfile, db=\"refseqplant\", evalue=0.001, outfmt='\"10 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore sgi sacc stitle\"', out=outputfile)\n",
      "#blastn_cline = NcbiblastnCommandline(query=inputfile, db=\"refseqplant\", evalue=0.001, outfmt=\"10 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore stitle\", out=outputfile)\n",
      "blastn_cline\n",
      "#can use std for all qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore\n",
      "#download one fasta seq from NCBI for a plant and blast it against this to see the difference between stitle and salltitles\n",
      "print(blastn_cline)\n",
      "stdout, stderr = blastn_cline()\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "blastn -out C:\\Users\\Alex\\Desktop\\Work\\Research\\Lettuce\\SampleBlast10.csv -outfmt \"10 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore sgi sacc stitle\" -query C:\\Users\\Alex\\Desktop\\Work\\Research\\Lettuce\\Samplesequence.txt -db refseqplant -evalue 0.001\n"
       ]
      }
     ],
     "prompt_number": 16
    },
    {
     "cell_type": "raw",
     "metadata": {},
     "source": [
      "blastn \u2013out C:\\Users\\Alex\\Desktop\\Work\\Research\\Lettuce\\SampleBlast10.csv -outfmt \"10 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore sgi sacc stitle\" \u2013query C:\\Users\\Alex\\Desktop\\Work\\Research\\Lettuce\\Samplesequence.txt -db refseqplant -evalue 0.001"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "TYPE THE ORGANISM NAME BELOW"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "DesiredOrganism = \"\""
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "import csv\n",
      "from Bio import Entrez\n",
      "Entrez.email = \"alexander.winnett001@umb.edu\"\n",
      "\n",
      "organisms = []\n",
      "\n",
      "with open(outputfile, 'rb') as csvfile: #open the CSV file generated by local blast\n",
      "     record = csv.reader(csvfile, delimiter=' ', quotechar='|') #name the records\n",
      "     for row in record: #for each entry in the spreadsheet\n",
      "            if row[10] == 0: #the E value must be 0\n",
      "                accession = row[12] #identify the accession number\n",
      "                handle = Entrez.efetch(db=\"nucleotide\", rettype=\"gb\", retmode=\"text\", id=accession) # db = database to query, rettype = retrieval type, retmode = retrieval mode \n",
      "                seq_record = SeqIO.read(handle, \"gb\") #read the contents \"fetched\" from NCBI\n",
      "                organism = seq_record[org] #\n",
      "                organisms.append(organism)\n",
      "                handle.close()\n",
      "                if organism in record[14]: #is the organism name is in the record\n",
      "                    orglen = len(organism) #findout how many letters the name of the organism is\n",
      "                    orgname = slice.organism(0:orglen)\n",
      "                    \n",
      "                \n",
      "                else:\n",
      "                    print \"Something went wrong\"\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "for record in SeqIO.parse(outputfile):\n",
      "    if record[10] == 0:\n",
      "        gi = record[12]\n",
      "        "
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "raw",
     "metadata": {},
     "source": [
      "OPEN CSV file\n",
      "look at ones with only E value of 0,\n",
      "then by organism too - but to use that, we'd have to have it search/slice for just the org name\n",
      "     to know what org names we can look from, we need the full list of organisms in the plant db\n",
      "        cut out the GeneID's so row# = gId --> use the gene ID to look up in NCBI, then from there, pull attribute from record that is = to organism\n",
      "        take that organism string and look for it in stitle (var = re.search(~~~, instring)) var.start() var.end() \n",
      "            ask if start is always at the beginning - if it is, then we can slice everything after the end position and denote as 'what it is'\n",
      "write the output to a new CSV file\n",
      "    contain all the information from the good hits, with extra column for organism, and gene name ('what it is')\n",
      "\n",
      "\n",
      "\n",
      "2. \n",
      "git hub\n",
      "    next project is similar to this, but has way more code files that call on each other, so keeping the versions straight will get tricky \n",
      "sign up for a github account - go through the tutorial"
     ]
    },
    {
     "cell_type": "raw",
     "metadata": {},
     "source": [
      "blastx -query opuntia.fasta -db nr -out opuntia.xml -evalue 0.001 -outfmt 10\n",
      "\n",
      "help(NcbiblastxCommandline)\n",
      "\n",
      "\n",
      "blastx_cline = NcbiblastxCommandline(query=\"opuntia.fasta\", db=\"nr\", evalue=0.001, outfmt=5, out=\"opuntia.xml\")\n",
      "blastx_cline\n",
      "NcbiblastxCommandline(cmd='blastx', out='opuntia.xml', outfmt=5, query='opuntia.fasta',\n",
      "db='nr', evalue=0.001)\n",
      "print(blastx_cline)\n",
      "blastx -out opuntia.xml -outfmt 5 -query opuntia.fasta -db nr -evalue 0.001 stdout, stderr = blastx_cline()"
     ]
    }
   ],
   "metadata": {}
  }
 ]
}