{
 "metadata": {
  "name": "",
  "signature": "sha256:b36b0b5a3efedf1ba6a314548f62b2282b642d424ec510e220e3f97d2784a776"
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "raw",
     "metadata": {},
     "source": [
      "/#inputfile = \"QGG24G02.yg.ab1_CLS_S3_Contig10858_LettuceNCBI.fasta\"\n",
      "#outputfile = inputfile+\"_\"+\"refseqBLASTresults.csv\"\n",
      "                   \n",
      "inputfile = \"C:\\Users\\Alex\\Desktop\\Work\\Research\\Lettuce\\\"\n",
      "outputfile = \"C:\\Users\\Alex\\Desktop\\Work\\Research\\Lettuce\\\"\n",
      "\n",
      "blastn_cline = NcbiblastnCommandline(query=inputfile, db=\"refseqplant\", evalue=0.001, outfmt='\"10 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore sgi sacc stitle\"', out=outputfile)\n",
      "#blastn_cline = NcbiblastnCommandline(query=inputfile, db=\"refseqplant\", evalue=0.001, outfmt=\"10 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore stitle\", out=outputfile)\n",
      "blastn_cline\n",
      "#can use std for all qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore\n",
      "#download one fasta seq from NCBI for a plant and blast it against this to see the difference between stitle and salltitles\n",
      "print(blastn_cline)\n",
      "stdout, stderr = blastn_cline()\n"
     ]
    },
    {
     "cell_type": "heading",
     "level": 6,
     "metadata": {},
     "source": [
      "blastn \u2013out C:\\PATH\\OUTPUTFILENAME.csv -outfmt \"10 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore sgi sacc stitle\" \u2013query C:\\PATH\\INPUTFILENAME.txt -db refseqplant -evalue 0.001"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "TYPE THE ORGANISM NAME BELOW"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "ename": "SyntaxError",
       "evalue": "invalid syntax (<ipython-input-44-023e7cc50b09>, line 1)",
       "output_type": "pyerr",
       "traceback": [
        "\u001b[1;36m  File \u001b[1;32m\"<ipython-input-44-023e7cc50b09>\"\u001b[1;36m, line \u001b[1;32m1\u001b[0m\n\u001b[1;33m    TYPE THE ORGANISM NAME BELOW\u001b[0m\n\u001b[1;37m           ^\u001b[0m\n\u001b[1;31mSyntaxError\u001b[0m\u001b[1;31m:\u001b[0m invalid syntax\n"
       ]
      }
     ],
     "prompt_number": 44
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "from Bio import SeqIO\n",
      "import csv\n",
      "from Bio import Entrez\n",
      "from Bio.Blast import NCBIWWW"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 45
    },
    {
     "cell_type": "code",
     "collapsed": true,
     "input": [
      "Entrez.email = \"alexander.winnett001@umb.edu\"\n",
      "\n",
      "\n",
      "localblastresults = \"C:\\Users\\Alex\\Desktop\\Work\\Research\\Lettuce\\GitHub\\Finding-EST-Sequences\\TESTLOCALBLAST_QGG24G02.yg.ab1_CLS_S3_Contig10858_LettuceNCBI.csv\"\n",
      "outputfile = \"C:\\Users\\Alex\\Desktop\\Work\\Research\\Lettuce\\GitHub\\Finding-EST-Sequences\\CORRECTEDTESTLOCALBLAST_QGG24G02.yg.ab1_CLS_S3_Contig10858_LettuceNCBI.csv\"\n",
      "\n",
      "organisms = []\n",
      "mismatch = 0\n",
      "duplicates = {}\n",
      "rows = 0 #limiter for testing\n",
      "with open(localblastresults, 'rb') as csvfile: #open the CSV file generated by local blast\n",
      "    record = csv.reader(csvfile)#name the records\n",
      "    #output = csv.writer(open(outputfile, 'wb'), delimiter=' ', quotechar='|', quoting=csv.QUOTE_MINIMAL)\n",
      "    #output.writerows(record)\n",
      "    for row in record: #for each entry in the spreadsheet\n",
      "        accession = row[12] #identify the accession number\n",
      "        handle = Entrez.efetch(db=\"nucleotide\", rettype=\"gb\", retmode=\"text\", id=accession) # db = database to query, rettype = retrieval type, retmode = retrieval mode \n",
      "        #seq_record = SeqIO.parse(handle, \"gb\") #read the contents \"fetched\" from NCBI\n",
      "        handle = Entrez.efetch(db=\"nucleotide\", rettype=\"gb\", retmode=\"text\", id=accession) # db = database to query, rettype = retrieval type, retmode = retrieval mode \n",
      "        seq_record = SeqIO.read(handle, \"gb\") #read the contents \"fetched\" from NCBI\n",
      "        org = seq_record.annotations['organism'] #pulls the organism name from the BLAST\n",
      "        stitle = row[14] #STITLE contains the org name and the name of the match\n",
      "        rows += 1  #for testing\n",
      "        print rows, \"rows analyzed\"\n",
      "        assert rows < 25 #LIMITER FOR TESTING#################################################\n",
      "        if org not in organisms:\n",
      "            organisms.append(org) #add that organism to the dictionary\n",
      "            print organisms\n",
      "            duplicates[org] = 0\n",
      "        elif org in organisms:\n",
      "            for name in duplicates:\n",
      "                if org == name:\n",
      "                    duplicatenumber = duplicates[org] + 1\n",
      "                    duplicates[org] = duplicatenumber\n",
      "            print \"duplicate found\"\n",
      "            duplicates[org] #keeping tabs of how many times an organism is hit, could be interesting\n",
      "        if org not in stitle: \n",
      "            mismatch += 1 #this is strange, and we should know how many times this occurs\n",
      "            print mismatch, \"mismatched\"\n",
      "        #output.writerow(row + [org])\n",
      "#I am having trouble appending the organism name onto the end of the new CSV file...\n",
      "        handle.close()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "1 rows analyzed\n",
        "['Ricinus communis']\n",
        "2"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        " rows analyzed\n",
        "['Ricinus communis', 'Vitis vinifera']\n",
        "3"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        " rows analyzed\n",
        "['Ricinus communis', 'Vitis vinifera', 'Populus trichocarpa']\n",
        "4"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        " rows analyzed\n",
        "duplicate found\n",
        "5"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        " rows analyzed\n",
        "duplicate found\n",
        "6"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        " rows analyzed\n",
        "duplicate found\n",
        "7"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        " rows analyzed\n",
        "['Ricinus communis', 'Vitis vinifera', 'Populus trichocarpa', 'Cicer arietinum']\n",
        "8"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        " rows analyzed\n",
        "duplicate found\n",
        "9"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        " rows analyzed\n",
        "duplicate found\n",
        "10"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        " rows analyzed\n",
        "duplicate found\n",
        "11"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        " rows analyzed\n",
        "duplicate found\n",
        "12"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        " rows analyzed\n",
        "duplicate found\n",
        "13"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        " rows analyzed\n",
        "['Ricinus communis', 'Vitis vinifera', 'Populus trichocarpa', 'Cicer arietinum', 'Eutrema salsugineum']\n",
        "14"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        " rows analyzed\n",
        "duplicate found\n",
        "15"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        " rows analyzed\n",
        "['Ricinus communis', 'Vitis vinifera', 'Populus trichocarpa', 'Cicer arietinum', 'Eutrema salsugineum', 'Phoenix dactylifera']\n",
        "16"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        " rows analyzed\n",
        "duplicate found\n",
        "17"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        " rows analyzed\n",
        "duplicate found\n",
        "18"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        " rows analyzed\n",
        "duplicate found\n",
        "19"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        " rows analyzed\n",
        "duplicate found\n",
        "20"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        " rows analyzed\n",
        "duplicate found\n",
        "21"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        " rows analyzed\n",
        "duplicate found\n",
        "22"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        " rows analyzed\n",
        "['Ricinus communis', 'Vitis vinifera', 'Populus trichocarpa', 'Cicer arietinum', 'Eutrema salsugineum', 'Phoenix dactylifera', 'Solanum lycopersicum']\n",
        "23"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        " rows analyzed\n",
        "['Ricinus communis', 'Vitis vinifera', 'Populus trichocarpa', 'Cicer arietinum', 'Eutrema salsugineum', 'Phoenix dactylifera', 'Solanum lycopersicum', 'Glycine max']\n",
        "24"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        " rows analyzed\n",
        "duplicate found\n",
        "25"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        " rows analyzed\n"
       ]
      },
      {
       "ename": "AssertionError",
       "evalue": "",
       "output_type": "pyerr",
       "traceback": [
        "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m\n\u001b[1;31mAssertionError\u001b[0m                            Traceback (most recent call last)",
        "\u001b[1;32m<ipython-input-145-37dd59401a02>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m()\u001b[0m\n\u001b[0;32m     23\u001b[0m         \u001b[0mrows\u001b[0m \u001b[1;33m+=\u001b[0m \u001b[1;36m1\u001b[0m  \u001b[1;31m#for testing\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     24\u001b[0m         \u001b[1;32mprint\u001b[0m \u001b[0mrows\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34m\"rows analyzed\"\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 25\u001b[1;33m         \u001b[1;32massert\u001b[0m \u001b[0mrows\u001b[0m \u001b[1;33m<\u001b[0m \u001b[1;36m25\u001b[0m \u001b[1;31m#LIMITER FOR TESTING#################################################\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     26\u001b[0m         \u001b[1;32mif\u001b[0m \u001b[0morg\u001b[0m \u001b[1;32mnot\u001b[0m \u001b[1;32min\u001b[0m \u001b[0morganisms\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     27\u001b[0m             \u001b[0morganisms\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0morg\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;31m#add that organism to the dictionary\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
        "\u001b[1;31mAssertionError\u001b[0m: "
       ]
      }
     ],
     "prompt_number": 145
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "print \"Organism Hits\"\n",
      "print organisms #contains all hit organisms only once\n",
      "print\n",
      "print \"Mismatches:\", \n",
      "print mismatch #number of times an org name isn't in stitle\n",
      "print \n",
      "print \"Duplication Trends\" \n",
      "print duplicates #dictionary detailing how many times each organism was a blast hit"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "Organism Hits\n",
        "['Ricinus communis', 'Vitis vinifera', 'Populus trichocarpa', 'Cicer arietinum', 'Eutrema salsugineum', 'Phoenix dactylifera', 'Solanum lycopersicum', 'Glycine max']\n",
        "\n",
        "Mismatches 0\n",
        "\n",
        "Duplication Trends\n",
        "{'Solanum lycopersicum': 0, 'Phoenix dactylifera': 3, 'Populus trichocarpa': 1, 'Cicer arietinum': 5, 'Glycine max': 0, 'Vitis vinifera': 3, 'Ricinus communis': 1, 'Eutrema salsugineum': 3}\n"
       ]
      }
     ],
     "prompt_number": 126
    },
    {
     "cell_type": "raw",
     "metadata": {},
     "source": [
      "OPEN CSV file\n",
      "look at ones with only E value of 0,\n",
      "then by organism too - but to use that, we'd have to have it search/slice for just the org name\n",
      "     to know what org names we can look from, we need the full list of organisms in the plant db\n",
      "        cut out the GeneID's so row# = gId --> use the gene ID to look up in NCBI, then from there, pull attribute from record that is = to organism\n",
      "        take that organism string and look for it in stitle (var = re.search(~~~, instring)) var.start() var.end() \n",
      "            ask if start is always at the beginning - if it is, then we can slice everything after the end position and denote as 'what it is'\n",
      "write the output to a new CSV file\n",
      "    contain all the information from the good hits, with extra column for organism, and gene name ('what it is')\n",
      "\n",
      "\n",
      "\n",
      "2. \n",
      "git hub\n",
      "    next project is similar to this, but has way more code files that call on each other, so keeping the versions straight will get tricky \n",
      "sign up for a github account - go through the tutorial"
     ]
    }
   ],
   "metadata": {}
  }
 ]
}