{
 "metadata": {
  "name": "",
  "signature": "sha256:abfbe6fff1d325f7a79c067990cee569c8042f6f864f32a9c50efeb65d70939c"
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "raw",
     "metadata": {},
     "source": [
      "#inputfile = \"QGG24G02.yg.ab1_CLS_S3_Contig10858_LettuceNCBI.fasta\"\n",
      "#outputfile = inputfile+\"_\"+\"refseqBLASTresults.csv\"\n",
      "                   \n",
      "inputfile = \"C:\\Users\\Alex\\Desktop\\Work\\Research\\Lettuce\\\"\n",
      "outputfile = \"C:\\Users\\Alex\\Desktop\\Work\\Research\\Lettuce\\\"\n",
      "\n",
      "blastn_cline = NcbiblastnCommandline(query=inputfile, db=\"refseqplant\", evalue=0.001, outfmt='\"10 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore sgi sacc stitle\"', out=outputfile)\n",
      "#blastn_cline = NcbiblastnCommandline(query=inputfile, db=\"refseqplant\", evalue=0.001, outfmt=\"10 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore stitle\", out=outputfile)\n",
      "blastn_cline\n",
      "#can use std for all qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore\n",
      "#download one fasta seq from NCBI for a plant and blast it against this to see the difference between stitle and salltitles\n",
      "print(blastn_cline)\n",
      "stdout, stderr = blastn_cline()\n"
     ]
    },
    {
     "cell_type": "heading",
     "level": 6,
     "metadata": {},
     "source": [
      "blastn \u2013out C:\\PATH\\OUTPUTFILENAME.csv -outfmt \"10 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore sgi sacc stitle\" \u2013query C:\\PATH\\INPUTFILENAME.txt -db refseqplant -evalue 0.001"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "TYPE THE ORGANISM NAME BELOW"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "from Bio import SeqIO\n",
      "import csv\n",
      "from Bio import Entrez\n",
      "from Bio.Blast import NCBIWWW"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 21
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "Entrez.email = \"alexander.winnett001@umb.edu\"\n",
      "\n",
      "\n",
      "localblastresults = \"C:\\Users\\Alex\\Desktop\\Work\\Research\\Lettuce\\GitHub\\Finding-EST-Sequences\\TESTLOCALBLAST_QGG24G02.yg.ab1_CLS_S3_Contig10858_LettuceNCBI.csv\"\n",
      "organisms = []\n",
      "\n",
      "with open(localblastresults, 'rb') as csvfile: #open the CSV file generated by local blast\n",
      "     record = csv.reader(csvfile)#name the records\n",
      "     for row in record: #for each entry in the spreadsheet\n",
      "        print len(row)#testing\n",
      "        print row[10]#testing\n",
      "        print row[12]#fortesting\n",
      "        print type(row[14])#for testing\n",
      "        accession = row[12] #identify the accession number\n",
      "        handle = Entrez.efetch(db=\"nucleotide\", rettype=\"gb\", retmode=\"text\", id=accession) # db = database to query, rettype = retrieval type, retmode = retrieval mode \n",
      "        seq_record = SeqIO.parse(handle, \"gb\") #read the contents \"fetched\" from NCBI\n",
      "        #handle = Entrez.efetch(db=\"nucleotide\", rettype=\"gb\", retmode=\"text\", id=accession) # db = database to query, rettype = retrieval type, retmode = retrieval mode \n",
      "        print handle\n",
      "        seq_record = SeqIO.read(handle, \"gb\") #read the contents \"fetched\" from NCBI\n",
      "        handle.close()\n",
      "        for k in seq_record.features:\n",
      "            organ = k.qualifiers['organism']#for testing\n",
      "            print organ[0]\n",
      "            if organ[0] not in organisms:\n",
      "                organisms.append(organ)\n",
      "            stitle = record[14]\n",
      "            if organ[0] in stitle: #is the organism name is in the record\n",
      "                orglen = len(organism) #findout how many letters the name of the organism is\n",
      "            print \"ITS HERE!\"\n",
      "            #orgname = slice.record[14](0:orglen)\n",
      "        else:\n",
      "            print \"Something went wrong\"\n",
      "        break\n",
      "        \n",
      "print organisms\n",
      "          "
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "17\n",
        "0.00\n",
        "255574376\n",
        "<type 'str'>\n",
        "<addinfourl at 67388872L whose fp = <socket._fileobject object at 0x0000000003EA9F48>>"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "\n",
        "Ricinus communis\n"
       ]
      },
      {
       "ename": "TypeError",
       "evalue": "'_csv.reader' object has no attribute '__getitem__'",
       "output_type": "pyerr",
       "traceback": [
        "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m\n\u001b[1;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
        "\u001b[1;32m<ipython-input-67-73d02ca1e198>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m()\u001b[0m\n\u001b[0;32m     24\u001b[0m             \u001b[1;32mif\u001b[0m \u001b[0morgan\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;36m0\u001b[0m\u001b[1;33m]\u001b[0m \u001b[1;32mnot\u001b[0m \u001b[1;32min\u001b[0m \u001b[0morganisms\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     25\u001b[0m                 \u001b[0morganisms\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0morgan\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 26\u001b[1;33m             \u001b[0mstitle\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mrecord\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;36m14\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     27\u001b[0m             \u001b[1;32mif\u001b[0m \u001b[0morgan\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;36m0\u001b[0m\u001b[1;33m]\u001b[0m \u001b[1;32min\u001b[0m \u001b[0mstitle\u001b[0m\u001b[1;33m:\u001b[0m \u001b[1;31m#is the organism name is in the record\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     28\u001b[0m                 \u001b[0morglen\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mlen\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0morganism\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;31m#findout how many letters the name of the organism is\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
        "\u001b[1;31mTypeError\u001b[0m: '_csv.reader' object has no attribute '__getitem__'"
       ]
      }
     ],
     "prompt_number": 67
    },
    {
     "cell_type": "raw",
     "metadata": {},
     "source": [
      "for record in SeqIO.parse(outputfile):\n",
      "    if record[10] == 0:\n",
      "        gi = record[12]\n",
      "        "
     ]
    },
    {
     "cell_type": "raw",
     "metadata": {},
     "source": [
      "OPEN CSV file\n",
      "look at ones with only E value of 0,\n",
      "then by organism too - but to use that, we'd have to have it search/slice for just the org name\n",
      "     to know what org names we can look from, we need the full list of organisms in the plant db\n",
      "        cut out the GeneID's so row# = gId --> use the gene ID to look up in NCBI, then from there, pull attribute from record that is = to organism\n",
      "        take that organism string and look for it in stitle (var = re.search(~~~, instring)) var.start() var.end() \n",
      "            ask if start is always at the beginning - if it is, then we can slice everything after the end position and denote as 'what it is'\n",
      "write the output to a new CSV file\n",
      "    contain all the information from the good hits, with extra column for organism, and gene name ('what it is')\n",
      "\n",
      "\n",
      "\n",
      "2. \n",
      "git hub\n",
      "    next project is similar to this, but has way more code files that call on each other, so keeping the versions straight will get tricky \n",
      "sign up for a github account - go through the tutorial"
     ]
    },
    {
     "cell_type": "raw",
     "metadata": {},
     "source": [
      "blastx -query opuntia.fasta -db nr -out opuntia.xml -evalue 0.001 -outfmt 10\n",
      "\n",
      "help(NcbiblastxCommandline)\n",
      "\n",
      "\n",
      "blastx_cline = NcbiblastxCommandline(query=\"opuntia.fasta\", db=\"nr\", evalue=0.001, outfmt=5, out=\"opuntia.xml\")\n",
      "blastx_cline\n",
      "NcbiblastxCommandline(cmd='blastx', out='opuntia.xml', outfmt=5, query='opuntia.fasta',\n",
      "db='nr', evalue=0.001)\n",
      "print(blastx_cline)\n",
      "blastx -out opuntia.xml -outfmt 5 -query opuntia.fasta -db nr -evalue 0.001 stdout, stderr = blastx_cline()"
     ]
    }
   ],
   "metadata": {}
  }
 ]
}